### ISSUE TITLE
Need to ban apparent direct eval.

### ISSUE THREAD

**Issue by erights** [Agoric/realms-shim#12]
The Realm shim already rejects -- by a cheap regexp test -- apparent dynamic import expressions. On the moduleplus branch at https://github.com/Agoric/realms-shim/blob/moduleplus/src/sourceParser.js we also reject html comments. @SMotaal just reminded me of the other thing the realms shim should reject with a cheap regexp test: apparent direct eval expressions.

We must ban dynamic imports and html comments for safety reasons. Therefore our regexp should err in only one direction: banning code that is actually safe. The motivation for banning direct eval is different. The realms proposal will specify support for direct eval. The realms shim cannot emulate direct eval. If allowed, it would execute with the semantics of indirect eval. If code accumulates which counts on that incorrect behavior, it could deter platforms from the providing the correct behavior later on. However, if our regexp does not successfully ban all apparent direct evals even in odd cases no one would write, for example `(eval)(str)`, this does not lead to unsafety. So our regexp can tolerate these slipping through.

**Comment by erights** [comment]
@SMotaal also suggested `eval(sanitize_for_eval(…))`, for some spelling of `sanitize_for_eval`, which is interesting. I explain the following engineering option in case it becomes needed. But I recommend not doing it unless desperate. Although it is likely safe, it skates too close to the edge.

We should not have the realms shim rewrite `eval(...)` to `eval(sanitize_for_eval(…))` because it would rewrite in comments and literal strings as well, and that's too invasive. This is why the shim either accepts or rejects but never rewrites.

However, the shim could have a special case that it rejects direct eval except for `eval(sanitize_for_eval(…))`, where it also checks that `sanitize_for_eval` (of some suitable spelling) does not appear anywhere else. Then, the module-to-static module record translation could accurately rewrite direct evals to that form, which the realms shim would then accept.

`sanitize_for_eval` would have to do something *extremely* evil and dangerous though, which is to arrange that the very next [[Get]] trap on the proxy, which must be for the name 'eval', would return the unsafe eval function itself, which the confined code must never get its hands on. This would reuse the delicate mechanism we've already reviewed for doing a direct eval safely within the magic 8 lines of code at 
https://github.com/Agoric/realms-shim/blob/master/src/evaluators.js#L59

**Comment by SMotaal** [comment]
You've effectively evolved an almost accidental suggestion into a well defined behaviour — this is just too good to not try out.

**Comment by erights** [comment]
The strategy for emulating direct eval explained above would not quite work, but is easily fixed. The problem is that `eval(sanitize_for_eval(…))` evaluates the lexical variable lookup of `eval` before it evaluates the rest. The evil strategy explained above relies on `sanitize_for_eval(…)` enabling the *next* faulting of the name `eval` to lookup the original unsafe eval function. But this doesn't work because the relevant lookup has already happened. Instead, rewriting to:

```js
($h_prepare_direct_eval(), eval(...))
```

or similar would work, for some spelling of `$h_prepare_direct_eval`, given that we also verify that `$h_prepare_direct_eval` does not occur anywhere else. The techniques can be combined to emulate as well the proposed Realms direct eval hook:

```js
($h_prepare_direct_eval(), eval($h_rewrite_for_direct_eval(...)))
```

Again, I want to emphasize that we do not anticipate any practical reason for the Realms or SES shim to risk leaking such dangerous powers in order to emulate direct eval. Rather, we should just banish direct eval until such a use case crops us. At that point, we can revisit the tradeoff.

But experimenting with this in the meantime should be fun! @SMotaal I'll be curious what happens. Thanks.

**Comment by erights** [comment]
In progress at https://github.com/Agoric/realms-shim/pull/13

**Comment by erights** [comment]
Fixes at #13
